src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java

```diff
@@ -385,21 +363,66 @@ private void processOffers() {
 
     @Override
     public synchronized void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
+        if (multiThreaded && !isProcessing()) {
+            startProcessing();
+        }
+
         for (Protos.Offer offer : offers) {
             boolean queued = offerQueue.offer(offer);
             if (!queued) {
                 LOGGER.warning("Offer queue is full.");
                 declineShort(offer);
+            } else {
+                LOGGER.info("Queued offer " + offer.getId().getValue());
             }
-
-            LOGGER.info("Queued offer " + offer.getId().getValue());
         }
 
         if (!multiThreaded) {
             processOffers();
         }
     }
 
+    @VisibleForTesting
+    String getFrameworkId() {
+        if (frameworkId != null) {
+            return frameworkId.getValue();
+        } else {
+            return NULL_FRAMEWORK_ID;
+        }
+    }
+
+    @VisibleForTesting
+    void startProcessing() {
+        String threadName = "mesos-offer-processor-" + getFrameworkId();
+        LOGGER.info("Starting offer processing thread: " + threadName);
+
+        offerProcessingThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                LOGGER.info("Started offer processing thread: " + threadName);
+                while(true) {
```

Comment:
Should probably add a catch-all `try`/`catch` around `processOffers()` that logs any exception thrown before the thread dies. Otherwise I think exceptions won't appear anywhere?

It could also make sense to have the catch-all just attempt to resume processing rather than killing the thread. But given the re-attempts to `startProcessing()` if the thread goes down, it should be fine to just `return;` from `run()` and let the thread stop running when an exception is thrown.